tests/test_settings.py

```diff
@@ -126,6 +128,53 @@ def test_base_path_overrides_config_file_code_and_deps(
     assert settings.deps == expected
 
 
+CODE_VALUES = ["a", "b", "c"]
+DEPS_VALUES = ["d", "e"]
+UNDECLARED_VALUES = ["f", "g"]
+UNUSED_VALUES = ["h", "i", "j"]
+
+
+def multivalued_optargs_grid():
+    def powerset(iterable):
+        xs = list(iterable)
+        return chain.from_iterable(combinations(xs, k) for k in range(len(xs) + 1))
+
+    keyed_items = {
+        "code": CODE_VALUES,
+        "deps": DEPS_VALUES,
+        "ignore-undeclared": UNDECLARED_VALUES,
+        "ignore-unused": UNUSED_VALUES,
+    }
+    rev_keyed = {v: k for k, vs in keyed_items.items() for v in vs}
+    ret = []
+    for items in keyed_items.values():
+        split_2 = set(
+            chain.from_iterable(
+                [(x, y), (y, x)]
+                for x, y in combinations(powerset(items), 2)
+                if len(x + y) == len(items) and not set(x) & set(y) and (x and y)
+            )
+        )
+        ret.append(split_2)
+    ret = list(product(*ret))
+    ret2 = []
+    for param_grid in ret:
+        xss = list(chain(*param_grid))
+        random.shuffle(xss)
+        optargs = list(chain(*[[f"--{rev_keyed[xs[0]]}"] + list(xs) for xs in xss]))
+        ret2.append(optargs)
+    return ret2
+
+
+@pytest.mark.parametrize("optargs", multivalued_optargs_grid())
+def test_multivalued_options_are_aggregated_correctly(optargs):
+    settings = run_build_settings(optargs)
+    assert settings.code == to_path_set(CODE_VALUES)
+    assert settings.deps == to_path_set(DEPS_VALUES)
+    assert settings.ignore_undeclared == set(UNDECLARED_VALUES)
+    assert settings.ignore_unused == set(UNUSED_VALUES)
```

Comment:
Can save a few lines by moving the `keyed_items` above the helper function.

Also, my brain is already too fried on a Friday afternoon to make sense of the logic in here. Can I ask for a docstring and some more descriptive variable names than `ret` and `ret2`? 😉 

```suggestion
OPTION_VALUES = {
    "code": ["a", "b", "c"],
    "deps": ["d", "e"],
    "ignore-undeclared": ["f", "g"],
    "ignore-unused": ["h", "i", "j"],
}


def multivalued_optargs_grid():
    """This could really use a docstring..."""
    def powerset(iterable):
        xs = list(iterable)
        return chain.from_iterable(combinations(xs, k) for k in range(len(xs) + 1))

    rev_keyed = {v: k for k, vs in OPTION_VALUES.items() for v in vs}
    ret = []  # 
    for items in OPTION_VALUES.values():
        split_2 = set(
            chain.from_iterable(
                [(x, y), (y, x)]
                for x, y in combinations(powerset(items), 2)
                if len(x + y) == len(items) and not set(x) & set(y) and (x and y)
            )
        )
        ret.append(split_2)
    ret = list(product(*ret))
    ret2 = []
    for param_grid in ret:
        xss = list(chain(*param_grid))
        random.shuffle(xss)
        optargs = list(chain(*[[f"--{rev_keyed[xs[0]]}"] + list(xs) for xs in xss]))
        ret2.append(optargs)
    return ret2


@pytest.mark.parametrize("optargs", multivalued_optargs_grid())
def test_multivalued_options_are_aggregated_correctly(optargs):
    settings = run_build_settings(optargs)
    assert settings.code == to_path_set(OPTION_VALUES["code"])
    assert settings.deps == to_path_set(OPTION_VALUES["deps"])
    assert settings.ignore_undeclared == set(OPTION_VALUES["ignore-undeclared"])
    assert settings.ignore_unused == set(OPTION_VALUES["ignore-unused"])
```

Reply:
Haha yeah absolutely, bad variable names and lack of comments by me. I pieced this together from tinkering around in a Jupyter REPL and neglected to do a pass back through for readability improvements. Will do.

Reply:
I saved some space here:
https://github.com/tweag/FawltyDeps/pull/247/commits/b4f3429e083e322707aec74c8c81072ebb181646
and eventually made some better names and added a docstring here:
https://github.com/tweag/FawltyDeps/pull/247/commits/1b3ad8ef8a60f693955f6497a966de6f22f2979d